src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,212 @@
+import assert from "assert";
+import { CoinGeckoPrice } from "../coingecko";
+import { Logger } from "../relayFeeCalculator"; // @todo: Relocate Logger to utils?
+
+export { Logger } from "../relayFeeCalculator"; // Permit adapters to import a common definition.
+export type TokenPrice = CoinGeckoPrice; // Temporary inversion; CoinGecko should source from here.
+
+// Represents valid source for spot prices (Across API, CoinGecko, ...)
+export interface PriceFeedAdapter {
+  readonly name: string;
+  getTokenPrice(address: string, currency: string, platform: string): Promise<TokenPrice>;
+  getTokenPrices(addresses: string[], currency: string, platform: string): Promise<TokenPrice[]>;
+}
```

Comment:
This interface does not define any support for historical price feeds. The existing Coingecko classes from UMAprotocol/sdk and from this package _do_ support historical price feeds, although I don't know if anyone uses it within Across.

Might be nice to define an optional method that supports retrieving historical prices; I presume only CoinGecko would end up implementing it, so the PriceClient would need to support that only a subset of PriceFeeds can actually support historical prices.

In the event that historical prices are supported, I don't envision implementing any support for caching, because it's such a bespoke use case that caching isn't likely to be used.

Reply:
> i think the paramers here are really specific to coingecko, which is wierd to me if you are creating an interface for getting prices. coingeckos api is set up for getting price based on an address, but i dont know if thats commonly supported for other sources? I would expect most to be based on a symbol.

Yeah it's definitely true that CoinGecko has influenced the interface a lot - also because the attributes of the interface can be seen in the Across API coingecko endpoint, and those are the only two endpoints that are supported here. I think it's a good point and has also influenced the "platform", which is a CoinGecko concept. Referencing by ChainID and then mapping that to "ethereum", ... for CoinGecko would be more generic.

One challenge that has always bothered me about looking up via a token symbol is that they're not guaranteed to be globally unique. CoinGecko handles this by having their own internal mapping, and I believe it would be possible to dynamically query the correct/intended token by looking up tokens that share a specific symbol, and then find some second parameter to differentiate them on. It definitely risks ending up with a less-certain result though - so the advantage of using an address is that the caller takes full responsibility for identifying the correct token.

Reply:
im not suggesting to remove the lookup by token address, simply saying that the interface should be explicit with the function name that it is doing a lookup by token address. when someone who knows nothing about coingecko goes to implement this interface, i dont think they would intuitively think this would be a lookup by token address

Reply:
Fair comment :) I've updated the PR with method names:

`getPriceByAddress()`
`getPricesByAddress()`